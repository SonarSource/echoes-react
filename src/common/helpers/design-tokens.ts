/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import designTokensColors from '~generated/design-tokens-colors.json';

export type DesignTokens = keyof typeof designTokensColors;

export type DesignTokensColors = {
  [K in DesignTokens]: K extends `echoes-color-${string}` ? K : never;
}[DesignTokens];
