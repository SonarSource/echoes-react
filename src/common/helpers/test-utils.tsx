/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { RenderOptions, render as rtlRender } from '@testing-library/react';
import userEvent, { Options as UserEventsOptions } from '@testing-library/user-event';
import React, { ComponentProps, PropsWithChildren } from 'react';
import { IntlProvider } from 'react-intl';
import { PropsWithLabels } from '~types/utils';
import { TooltipProvider } from '../../components/tooltip';

export function render(
  ui: React.ReactElement,
  options?: RenderOptions,
  userEventOptions?: UserEventsOptions,
) {
  return {
    ...rtlRender(ui, { wrapper: ContextWrapper, ...options }),
    user: userEvent.setup(userEventOptions),
  };
}

function ContextWrapper({ children }: PropsWithChildren<{}>) {
  return (
    <IntlProvider defaultLocale="en-us" locale="en-us">
      <TooltipProvider delayDuration={0}>{children}</TooltipProvider>
    </IntlProvider>
  );
}

export type OmitPropsWithLabels<T extends React.JSXElementConstructor<any>> = Partial<
  Omit<ComponentProps<T>, keyof PropsWithLabels<{}>>
> &
  PropsWithLabels<{}>;
