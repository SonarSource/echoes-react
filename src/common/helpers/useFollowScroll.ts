/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { useEffect, useState } from 'react';
import { throttle, THROTTLE_SHORT_DELAY } from './utils';

/**
 * A React hook that tracks the scroll position of the document.
 * It returns an object containing the current scroll position (left and top).
 * The scroll position is updated at a throttled rate to improve performance.
 *
 * @returns An object with `left` and `top` properties representing the current scroll position.
 */
export function useFollowScroll(enabled = true) {
  const [left, setLeft] = useState(0);
  const [top, setTop] = useState(0);

  useEffect(() => {
    const followScroll = throttle(() => {
      if (document.documentElement) {
        setLeft(document.documentElement.scrollLeft);
        setTop(document.documentElement.scrollTop);
      }
    }, THROTTLE_SHORT_DELAY);

    if (enabled) {
      document.addEventListener('scroll', followScroll);
    }

    return () => document.removeEventListener('scroll', followScroll);
  }, [enabled]);

  return { left, top };
}
