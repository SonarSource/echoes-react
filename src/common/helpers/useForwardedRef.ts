/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { type ForwardedRef, useState } from 'react';

/**
 * This hook may be used to intercept a forwarded ref, providing a local ref
 * that can be used in the component.
 *
 * **Example**
 *
 * ```typescript
 * const Input = forwardRef<HTMLInputElement>((props, forwardedRef) => {
 *   const [ref, setRef] = useForwardedRef(forwardedRef);
 *   return <input ref={setRef} />
 * });
 * ```
 */
export function useForwardedRef<T>(forwardedRef: ForwardedRef<T>) {
  const [ref, setRef] = useState<T | null>(null);

  const setForwardedRef = (element: T | null) => {
    if (typeof forwardedRef === 'function') {
      forwardedRef(element);
    } else if (forwardedRef) {
      forwardedRef.current = element;
    }

    setRef(element);
  };

  return [ref, setForwardedRef] as const;
}
