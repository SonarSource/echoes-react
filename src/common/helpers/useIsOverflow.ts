/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { DependencyList, MutableRefObject, useLayoutEffect, useState } from 'react';

/**
 * This hook checks if the content of a given ref is overflowing its container horizontally.
 * It returns a boolean indicating whether the content is overflowing.
 *
 * @param forwardedRef - The ref to check for overflow.
 * @param deps - Optional dependencies array to re-compute the overflow state.
 */
export function useIsOverflow(
  forwardedRef: MutableRefObject<HTMLElement | null>,
  deps: DependencyList = [],
) {
  const [isOverflow, setIsOverflow] = useState<boolean | undefined>(undefined);

  useLayoutEffect(() => {
    if (forwardedRef.current) {
      setIsOverflow(forwardedRef.current.scrollWidth > forwardedRef.current.offsetWidth);
    }
    // We want to be able to pass an array of dependencies to re-compute the overflow state
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [forwardedRef, ...deps]);

  return [isOverflow] as const;
}
