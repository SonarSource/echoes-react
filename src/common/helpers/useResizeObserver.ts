/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { RefObject, useCallback, useEffect, useState } from 'react';

export function useResizeObserver(ref: RefObject<HTMLElement>) {
  const [width, setWidth] = useState<number>();
  const [height, setHeight] = useState<number>();

  const handleResize = useCallback<ResizeObserverCallback>((entries) => {
    if (!Array.isArray(entries)) {
      return;
    }

    const entry = entries[0];
    setWidth(entry.contentRect.width);
    setHeight(entry.contentRect.height);
  }, []);

  useEffect(() => {
    if (!ref.current) {
      return undefined;
    }

    const RO = new ResizeObserver(handleResize);
    RO.observe(ref.current);

    return () => {
      RO.disconnect();
    };
  }, [handleResize, ref]);

  return { width, height };
}
