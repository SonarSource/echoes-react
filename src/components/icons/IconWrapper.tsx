/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { isStringDefined } from '~common/helpers/types';

export interface IconProps {
  ariaLabel?: string;
  className?: string;
  /* To discuss?
  ['data-guiding-id']?: string;
  fill?: ThemeColors | CSSColor;
  style?: React.CSSProperties;
  */
}

export interface IconFilledProps extends IconProps {
  isFilled?: boolean;
}

type IconWrapperProps = JSX.IntrinsicElements['span'];

function IconBase(props: Readonly<IconWrapperProps & IconFilledProps>) {
  const {
    ariaLabel,
    'aria-label': nativeAriaLabel,
    children,
    isFilled, // extracted here to prevent it from leaking to the span element
    ...rest
  } = props;

  return (
    <span
      // 'aria-label' is not a valid prop, but it might be used by the consumer when wrapping the component with emotion
      aria-hidden={isStringDefined(ariaLabel ?? nativeAriaLabel) ? 'false' : 'true'}
      aria-label={ariaLabel ?? nativeAriaLabel}
      role="img"
      {...rest}>
      {children}
    </span>
  );
}

export const IconMaterialWrapper = styled(IconBase)<IconWrapperProps & IconFilledProps>`
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  font-family: 'Material Symbols Rounded';
  font-size: calc(1em + 4px);
  font-style: normal;
  font-weight: normal;
  height: calc(2em - 16px);
  line-height: calc(2em - 16px);
  text-align: center;
  vertical-align: bottom;
  width: calc(2em - 16px);

  ${({ isFilled = false }) => (isFilled ? `font-variation-settings: 'FILL' 1;` : '')}
`;
IconMaterialWrapper.displayName = 'IconMaterialWrapper';

export const IconCustomWrapper = styled(IconMaterialWrapper)<IconWrapperProps & IconProps>`
  font-family: 'Echoes';
`;
IconMaterialWrapper.displayName = 'IconCustomWrapper';
