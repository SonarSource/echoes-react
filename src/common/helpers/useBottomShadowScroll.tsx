/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { RefObject, useCallback, useEffect, useState } from 'react';
import { cssVar } from '~utils/design-tokens';
import { useResizeObserver } from './useResizeObserver';

/**
 * Custom hook to manage the visibility of a bottom shadow on a scrollable container.
 * The shadow indicates that there is more content to scroll to.
 * @param scrollableContainerRef - Ref to the scrollable container element, its height and scroll event are monitored to recompute the shadow
 * @param resizableContentRef - Ref to the resizable content element inside the scrollable container, its height is monitored to recompute the shadow
 * @returns {boolean} Whether the bottom shadow should be shown
 */
export function useBottomShadowScroll(
  scrollableContainerRef: RefObject<HTMLElement>,
  resizableContentRef: RefObject<HTMLElement>,
) {
  const [showBottomShadow, setShowBottomShadow] = useState(false);

  // Monitor the height of the scrollable container, and recompute the shadow if it changes
  const { height: scrollableContainerHeight } = useResizeObserver(scrollableContainerRef);
  // Monitor the height of the resizable content, and recompute the shadow if it changes, because it may move the scroll position
  const { height: resizableContentHeight } = useResizeObserver(resizableContentRef);

  const updateBottomShadowScroll = useCallback(() => {
    if (scrollableContainerRef.current) {
      const { scrollHeight, clientHeight, scrollTop } = scrollableContainerRef.current;
      setShowBottomShadow(scrollTop + clientHeight < scrollHeight - 8); // -8px for bottom padding
    }
  }, [scrollableContainerRef]);

  // Compute the bottom shadow visibility on first render and attach the scroll event listener
  useEffect(() => {
    const scrollableContainerRefNode = scrollableContainerRef.current;
    if (!scrollableContainerRefNode) {
      return undefined;
    }

    // Re-compute the bottom shadow visibility on scroll event
    scrollableContainerRefNode.addEventListener('scroll', updateBottomShadowScroll);

    // Initial computation
    updateBottomShadowScroll();

    return () => {
      scrollableContainerRefNode.removeEventListener('scroll', updateBottomShadowScroll);
    };
  }, [scrollableContainerRef, updateBottomShadowScroll]);

  // Re-compute the bottom shadow visibility on container and content resize
  useEffect(updateBottomShadowScroll, [
    resizableContentHeight,
    scrollableContainerHeight,
    updateBottomShadowScroll,
  ]);

  return [showBottomShadow] as const;
}

export const BottomShadowScroll = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;

  min-height: ${cssVar('dimension-height-400')};

  border-bottom: ${cssVar('border-width-default')} solid ${cssVar('color-border-weak')};
  background: radial-gradient(
      farthest-side at 50% 100%,
      ${cssVar('bottom-scroll-colors-shadow-gradient')},
      ${cssVar('color-support-transparent')}
    )
    center bottom;
`;
BottomShadowScroll.displayName = 'BottomShadowScroll';
