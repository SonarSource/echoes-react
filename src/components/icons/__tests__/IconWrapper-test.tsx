/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { screen } from '@testing-library/react';
import { render } from '../../../common/helpers/test-utils';
import { IconCustomWrapper, IconMaterialWrapper } from '../IconWrapper';

it('should render material icon correctly', () => {
  const { container } = render(<IconMaterialWrapper>&#xE3A6;</IconMaterialWrapper>);

  expect(screen.queryByRole('img')).not.toBeInTheDocument();

  const icon = screen.getByRole('img', { hidden: true });
  expect(icon).toContainHTML('&#xE3A6;');
  expect(icon).toHaveStyle({
    'font-family': `'Material Symbols Rounded'`,
  });
  expect(icon).not.toHaveStyle({
    'font-variation-settings': `'FILL' 1`,
    'font-family': `'Echoes'`,
  });
  expect(container).toMatchSnapshot();
});

it('should render custom icon correctly', () => {
  render(<IconCustomWrapper>&#x21;</IconCustomWrapper>);

  expect(screen.queryByRole('img')).not.toBeInTheDocument();

  const icon = screen.getByRole('img', { hidden: true });
  expect(icon).toContainHTML('&#x21;');
  expect(icon).not.toHaveStyle({
    'font-family': `'Material Symbols Rounded'`,
  });
  expect(icon).toHaveStyle({
    'font-family': `'Echoes'`,
  });
});

it('should not be hidden when ariaLabel is set', () => {
  render(<IconMaterialWrapper ariaLabel="test">&#xE3A6;</IconMaterialWrapper>);

  expect(screen.getByRole('img')).toBeVisible();
});

it('should correctly handled isFilled props for material icons', () => {
  render(<IconMaterialWrapper isFilled>&#xE3A6;</IconMaterialWrapper>);

  expect(screen.getByRole('img', { hidden: true })).toHaveStyle({
    'font-variation-settings': `'FILL' 1`,
  });
});

it('should accept custom color', () => {
  render(<IconMaterialWrapper color="--echoes-color-icon-success">&#xE3A6;</IconMaterialWrapper>);

  expect(screen.getByRole('img', { hidden: true })).toHaveStyle({
    color: `var(--echoes-color-icon-success)`,
  });
});
