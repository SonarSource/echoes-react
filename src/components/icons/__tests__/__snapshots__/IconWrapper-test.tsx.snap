// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render custom icon correctly 1`] = `
.emotion-0 {
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  font-family: 'Material Symbols Rounded';
  font-size: calc(1em + 4px);
  font-style: normal;
  font-weight: normal;
  height: calc(2em - 16px);
  line-height: calc(2em - 16px);
  text-align: center;
  vertical-align: bottom;
  width: calc(2em - 16px);
  font-family: 'Echoes';
}

<div>
  <span
    aria-hidden="true"
    class="emotion-0"
    role="img"
  >
    !
  </span>
</div>
`;

exports[`should render material icon correctly 1`] = `
.emotion-0 {
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  font-family: 'Material Symbols Rounded';
  font-size: calc(1em + 4px);
  font-style: normal;
  font-weight: normal;
  height: calc(2em - 16px);
  line-height: calc(2em - 16px);
  text-align: center;
  vertical-align: bottom;
  width: calc(2em - 16px);
}

<div>
  <span
    aria-hidden="true"
    class="emotion-0"
    role="img"
  >
    îŽ¦
  </span>
</div>
`;
